### PR TITLE
Improve data reading performance of partition table

### PR DESCRIPTION
**Pull request by qingxinhome**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #8320

## What this PR does / why we need it:
1. Open the partition table function
2. Improve data reading performance of partition table

### CODE REVIEW

**Review by reusee — APPROVED** [review]

**Comment by matrix-meow** [comment]
@qingxinhome Thanks for your contributions!

**Here are review comments for file pkg/frontend/compiler_context.go:**
The pull request aims to improve the data reading performance of the partition table. The changes made in the pull request are in the file `pkg/frontend/compiler_context.go`. The changes include opening the partition table function and improving the data reading performance of the partition table. The changes include modifying the `Stats` function to read data from the partition table. The changes include adding a loop to iterate over the engine definitions and extract partition information. The changes also include creating an array of partition tables and passing it to the `Stats` function. 

However, the pull request lacks a detailed explanation of the changes made. The pull request does not provide any information on how the changes improve the data reading performance of the partition table. The pull request also does not provide any information on how the changes affect the existing codebase. The pull request does not provide any information on how the changes affect the performance of other functions that use the `Stats` function. 

To improve the pull request, the author should provide a detailed explanation of the changes made. The author should explain how the changes improve the data reading performance of the partition table. The author should also explain how the changes affect the existing codebase. The author should provide information on how the changes affect the performance of other functions that use the `Stats` function. Additionally, the author should consider optimizing the changes made in the pull request to improve the overall performance of the codebase.

**Here are review comments for file pkg/frontend/test/engine_mock.go:**
The title of the pull request is clear and concise, indicating that the purpose of the changes is to improve the data reading performance of a partition table. The body of the pull request includes information about the type of pull request, the issue it fixes, and what the changes do and why they are needed. However, the body could be improved by providing more specific details about how the data reading performance is improved.

The changes made in the pull request are focused on the file `pkg/frontend/test/engine_mock.go`. The changes involve modifying the `Stats` method of the `MockStatistics` and `MockRelation` structs to include a new parameter `partitionTables []any`. However, the changes made in the pull request are incomplete and do not provide enough context to fully understand the purpose of the changes. 

Suggestions for improvement:
- Provide more specific details about how the data reading performance is improved in the body of the pull request.
- Provide more context about the changes made in the pull request, including why the new parameter `partitionTables []any` is needed and how it is used.
- Ensure that the changes made in the pull request are complete and do not introduce any new issues or errors.

**Here are review comments for file pkg/sql/colexec/constraint_util.go:**
The pull request aims to improve the data reading performance of the partition table. The changes made in the pull request are in the file `pkg/sql/colexec/constraint_util.go`. The changes involve commenting out the creation of a new vector and instead using the `proc.GetVector()` method to get a vector. 

There are a few issues with this pull request that need addressing:

1. The pull request title is not descriptive enough. The title should clearly state what the pull request is about. A better title would be "Improve data reading performance of partition table by optimizing vector creation".

2. The pull request body is incomplete. The body should explain in detail what changes were made and why they were made. It should also provide information on how the changes were tested. A better body would be:

   ```
   ## What type of PR is this?

   - [ ] API-change
   - [ ] BUG
   - [x] Improvement
   - [ ] Documentation
   - [ ] Feature
   - [ ] Test and CI
   - [ ] Code Refactoring

   ## Which issue(s) this PR fixes:

   issue #8320

   ## What this PR does / why we need it:

   This pull request aims to improve the data reading performance of the partition table by optimizing vector creation. The changes involve commenting out the creation of a new vector and instead using the `proc.GetVector()` method to get a vector. This change reduces the overhead of vector creation and improves the performance of the partition table.

   ## How to test:

   The changes were tested by running the existing test suite and measuring the performance of the partition table before and after the changes were made. The tests passed, and the performance of the partition table improved significantly.
   ```

3. The changes made in the pull request are not optimized. While using `proc.GetVector()` instead of creating a new vector reduces the overhead of vector creation, it still involves a function call. A better approach would be to reuse existing vectors instead of creating new ones. This would further reduce the overhead and improve performance.

To optimize the changes made in the pull request, the following changes can be made:

- Instead of creating a new vector or using `proc.GetVector()`, reuse existing vectors. This can be done by maintaining a pool of vectors that can be reused when needed. This approach would further reduce the overhead of vector creation and improve performance.

- Add benchmarks to measure the performance of the partition table before and after the changes are made. This would help identify any performance regressions and ensure that the changes made in the pull request are indeed improving performance.

**Here are review comments for file pkg/sql/compile/scope.go:**
The pull request aims to improve the data reading performance of a partition table. The changes made in the pull request are in the file `pkg/sql/compile/scope.go`. The changes include opening the partition table function and improving the data reading performance of the partition table. 

There are a few issues with this pull request that need to be addressed:

1. The title of the pull request is not descriptive enough. It should provide more information about what the pull request does and what changes it makes.

2. The body of the pull request is incomplete. It should provide more information about how the data reading performance of the partition table is improved.

3. The changes made in the pull request are not optimized. The added log statement can impact the performance of the code and should be removed before merging the pull request.

Suggestions to address these issues are:

1. The title of the pull request should be more descriptive. For example, "Improve data reading performance of partition table by optimizing scope.go".

2. The body of the pull request should provide more information about how the data reading performance of the partition table is improved. This can include details about the changes made to the code and how they impact the performance.

3. The added log statement should be removed before merging the pull request. This can be done by deleting the line `logutil.Infof("---->ParallelRun---> %s \n", DebugShowScopes([]*Scope{s}))`.

**Here are review comments for file pkg/sql/plan/build_ddl.go:**
The title of the pull request is clear and concise, indicating that the goal of the changes is to improve the data reading performance of a partition table. The body of the pull request provides some additional context, including the issue number that the changes address and a brief summary of the changes made.

The changes themselves appear to be focused on the `addPartitionTableDef` function in the `build_ddl.go` file. The changes involve uncommenting some code that was previously commented out and making some modifications to the code. Specifically, the changes involve adding some common properties to the partition table, constructing the table definition for the partition table, and saving the table name for each partition.

One potential problem with the pull request is that it is not entirely clear how the changes will improve the data reading performance of the partition table. The pull request title and body suggest that this is the goal, but the changes themselves do not seem to directly address this issue. It may be helpful for the author of the pull request to provide more information about how the changes will improve performance.

Another potential issue is that the code that was previously commented out has been uncommented without any explanation as to why it was commented out in the first place. It may be helpful for the author of the pull request to provide some context about why this code was commented out and why it is now being uncommented.

In terms of optimization, it is difficult to say without more information about the performance issues being addressed. However, it may be helpful for the author of the pull request to provide some benchmarks or other performance metrics to demonstrate the effectiveness of the changes.

Overall, the changes in the pull request appear to be focused on improving the partition table functionality, but more information is needed to fully understand the impact of the changes on performance. Additionally, some additional context about the uncommented code and some performance metrics would be helpful.

**Here are review comments for file pkg/sql/plan/partition_hash.go:**
The pull request title is clear and concise, indicating that it aims to improve the data reading performance of a partition table. The body of the pull request provides some additional context, such as the fact that it addresses issue #8320 and that it opens the partition table function. However, it could benefit from a more detailed explanation of how the changes made in the pull request improve performance.

The changes made in the pull request are minimal and only involve uncommenting a line of code. It is unclear how this change will improve performance, and it seems unlikely that uncommenting a single line of code would have a significant impact on performance. Therefore, it would be helpful if the pull request author provided more information on how this change improves performance.

Additionally, the pull request author did not provide any information on how they tested the changes made in the pull request. It would be helpful if they included information on how they tested the changes and what tests they ran to ensure that the changes did not introduce any new bugs or issues.

To optimize the changes made in the pull request, the pull request author could provide more detailed information on how the changes improve performance. They could also include information on how they tested the changes and what tests they ran to ensure that the changes did not introduce any new bugs or issues.

**Here are review comments for file pkg/sql/plan/partition_key.go:**
The title of the pull request is clear and concise, indicating that the goal of the changes is to improve the data reading performance of a partition table. However, the body of the pull request is lacking in detail. It only includes a checklist of the type of PR and the issue number that the changes address. It would be helpful to include more information about the specific changes made and how they improve performance.

The changes made in the pull request are minor and do not seem to directly address the stated goal of improving data reading performance. The only change made is to uncomment a line of code that sets the partition definition for a table. This change does not seem to have any impact on performance and may have been accidentally included in the pull request.

To optimize the changes made in the pull request, the author should provide more detailed information about the specific changes made and how they improve performance. Additionally, they should review the changes to ensure that they are relevant to the stated goal and do not include any extraneous or unrelated changes.

**Here are review comments for file pkg/sql/plan/partition_list.go:**
The pull request title is clear and concise, indicating that it aims to improve the data reading performance of the partition table. The body of the pull request provides additional information, including the type of PR, the issue it fixes, and what the PR does and why it is needed. However, the body could benefit from more detailed information about how the data reading performance was improved.

The changes made in the pull request are minimal and only involve uncommenting a line of code. It is unclear how this change improves data reading performance, and there are no additional changes or explanations provided in the pull request. Therefore, the pull request needs more information about how the uncommented line of code improves data reading performance. 

To optimize the changes made in the pull request, the author could provide more detailed information about the changes made and how they improve data reading performance. Additionally, the author could consider including performance benchmarks or tests to demonstrate the effectiveness of the changes.

**Here are review comments for file pkg/sql/plan/partition_range.go:**
The pull request title and body are clear and concise. However, the changes made in the pull request seem unrelated to the title and body. The changes only include a single line of code that is uncommented. It is unclear how this change relates to the stated goal of improving data reading performance of the partition table. 

Suggestions:
- The pull request should include more detailed information about the changes made and how they relate to the stated goal of improving data reading performance.
- The changes made should be more significant and related to the stated goal. 
- The pull request should include comments explaining the changes made and how they relate to the overall codebase.

**Here are review comments for file pkg/vm/engine/disttae/txn_table.go:**
The pull request aims to improve the data reading performance of the partition table. The changes made in the pull request are in the `pkg/vm/engine/disttae/txn_table.go` file. The `Stats` function has been modified to accept an additional parameter `partitionTables []any`. The function now iterates over the partition tables and calculates the sum of blockInfos. If the sum of blockInfos is greater than zero, it calls the `CalcStats` function to calculate the statistics. Otherwise, it returns false.

However, there are a few issues with this pull request that need to be addressed:

1. The title of the pull request is not descriptive enough. It should provide more information about the changes made in the pull request.

2. The body of the pull request does not provide enough information about the changes made in the pull request. It should include a detailed explanation of the changes made and why they are necessary.

3. The code changes made in the pull request are not optimized. The `CalcStats` function is called twice, which can be avoided by moving the code outside the if-else block.

4. The panic statement in the code is not a good practice. Instead of panicking, it should return an error message.

To address these issues, the pull request should be updated as follows:

1. The title of the pull request should be changed to "Improve data reading performance of partition table by summing blockInfos".

2. The body of the pull request should be updated to include a detailed explanation of the changes made and why they are necessary. It should also include information about the optimization made in the code.

3. The code changes made in the pull request should be optimized by moving the `CalcStats` function outside the if-else block.

4. The panic statement in the code should be replaced with an error message.

**Here are review comments for file pkg/vm/engine/memoryengine/table.go:**
The pull request title and body are clear and concise. However, the changes made in the pull request are not related to the title and body. The changes only modify the `Stats` function by adding a new parameter `partitionTables []any`. It is not clear how this change improves the data reading performance of the partition table. 

Suggestions:
- The pull request title and body should accurately reflect the changes made in the code.
- The pull request should provide a clear explanation of how the changes improve the data reading performance of the partition table.
- The changes made in the pull request should be related to the pull request title and body. 

Optimization:
- If the changes made in the pull request are related to improving data reading performance, then the pull request should include additional changes that address this issue.

**Here are review comments for file pkg/vm/engine/types.go:**
The title of the pull request is clear and concise, indicating that the purpose of the changes is to improve the data reading performance of a partition table. However, the body of the pull request is lacking in detail and context. It only includes a checklist of the type of PR and the issue number that it fixes. It would be helpful to include a brief explanation of the changes made and how they improve the performance of the partition table.

The changes made in the pull request involve modifying the `Stats` function in the `Statistics` interface in the `types.go` file. The function now takes an additional parameter, `partitionTables`, which is an array of any type. However, the use of `any` type is not recommended as it can lead to type errors and make the code harder to read and maintain. It would be better to use a more specific type or interface.

Additionally, it is not clear how the changes made in this pull request will improve the data reading performance of the partition table. It would be helpful to include more details on the specific optimizations made and how they will impact the performance of the partition table.

To optimize the changes made in the pull request, it would be helpful to include benchmarks or performance tests to demonstrate the improvements made. This would provide a clear and objective measure of the impact of the changes on the performance of the partition table.

**Here are review comments for file test/distributed/cases/ddl/partition3.sql:**
The pull request aims to improve the data reading performance of partition tables. The changes made in the pull request are in the file `test/distributed/cases/ddl/partition3.sql`. The changes include removing a comment and a newline character at the end of the file.

There are a few issues with this pull request that need to be addressed:

1. The title of the pull request is not descriptive enough. It should include more information about what changes were made and why they were made. A more descriptive title could be "Optimize data reading performance of partition table by removing unnecessary comments and characters".

2. The body of the pull request is also not detailed enough. It only provides a brief overview of the changes made and does not explain why they were necessary. The body should include more information about the performance issues that were identified and how the changes made will improve performance.

3. The changes made in the pull request are minimal and do not address the root cause of the performance issues. While removing unnecessary comments and characters can improve performance slightly, it is not a significant improvement. The pull request should include more substantial changes that address the root cause of the performance issues.

4. The changes made in the pull request do not include any tests to ensure that the changes do not introduce any new issues. It is important to include tests to ensure that the changes made do not break any existing functionality.

To optimize the changes made in the pull request, the following suggestions can be considered:

1. Include more detailed information in the title and body of the pull request to provide a clear understanding of the changes made and why they were necessary.

2. Address the root cause of the performance issues by making more substantial changes to the code.

3. Include tests to ensure that the changes made do not break any existing functionality.

By addressing these issues and implementing the suggested optimizations, the pull request can be improved to provide a more significant improvement in performance and ensure that the changes made do not introduce any new issues.

**Here are review comments for file test/distributed/cases/ddl/partition4.sql:**
The pull request aims to improve the data reading performance of a partition table. The changes made in the pull request are to the file `test/distributed/cases/ddl/partition4.sql`. The changes include removing a line that was commented out and adding a line to select data from a partition. 

The pull request's title and body are clear and concise, and they provide enough information about the changes made. However, there are a few issues that need to be addressed:

1. The pull request does not provide any details about how the data reading performance was improved. It would be helpful to have more information about the changes made to the partition table function to improve performance.

2. The pull request does not provide any information about how the changes were tested. It would be helpful to know if any performance tests were run to verify that the changes made actually improved performance.

3. The changes made to the `partition4.sql` file seem unrelated to the improvements made to the partition table function. It would be helpful to have more information about why these changes were made and how they relate to the improvements made to the partition table function.

To optimize the changes made in the pull request, the following suggestions can be considered:

1. Provide more details about how the data reading performance was improved. This could include information about any optimizations made to the partition table function, such as changes to the data structures used or improvements to the algorithm used to read data from partitions.

2. Provide information about how the changes were tested. This could include details about any performance tests that were run to verify that the changes made actually improved performance.

3. Clarify the changes made to the `partition4.sql` file and how they relate to the improvements made to the partition table function. This could include information about how the changes made to the file were used to test the improvements made to the partition table function.

**Here are review comments for file test/distributed/cases/ddl/table_partition.test:**
The pull request aims to improve the data reading performance of a partition table. The changes made in the pull request include opening the partition table function and modifying the file test/distributed/cases/ddl/table_partition.test. The modifications in the file include removing a line and adding another line. 

However, the pull request lacks detailed information on how the performance improvement was achieved. It would be helpful to provide more information on the specific changes made to the partition table function and how they improve performance. Additionally, the modifications made to the test file seem unrelated to the performance improvement and require further explanation.

Suggestions:
- Provide more detailed information on how the performance improvement was achieved.
- Explain the modifications made to the test file and how they relate to the performance improvement.
- Consider adding performance benchmarks to demonstrate the improvement achieved.

**Here are review comments for file test/distributed/cases/transaction/atomicity_1.sql:**
The pull request aims to improve the data reading performance of a partition table. The changes made in the pull request are only related to a single file `test/distributed/cases/transaction/atomicity_1.sql`. The changes include removing a comment and whitespace from the file. However, it is not clear how these changes are related to the performance improvement of the partition table.

Suggestions:
- The pull request title should be more descriptive and provide more information about the changes made in the pull request.
- The pull request body should provide more details about the performance improvement and how the changes made in the file are related to it.
- The changes made in the file should be explained more clearly and how they relate to the performance improvement.
- It would be helpful to include some performance metrics or benchmarks to demonstrate the improvement in data reading performance.

**Here are review comments for file test/distributed/cases/zz_accesscontrol/grant_privs_role.sql:**
The title of the pull request is clear and concise, indicating that the goal is to improve the data reading performance of a partition table. The body of the pull request includes a checklist of the type of PR, the issue number that the PR fixes, and a brief description of the changes made. However, the description is not detailed enough to provide a clear understanding of what specific changes were made to improve the data reading performance.

The changes made in the pull request are focused on a single file, `grant_privs_role.sql`, which is located in the `test/distributed/cases/zz_accesscontrol/` directory. The changes include removing two lines of code that reference an issue number and updating the `grant_privs_role.sql` file to improve data reading performance of a partition table. However, it is unclear how the changes made to the `grant_privs_role.sql` file improve data reading performance.

To improve the pull request, the description of the changes made should be more detailed and provide a clear explanation of how the changes improve data reading performance. Additionally, the changes made to the `grant_privs_role.sql` file should be explained in more detail, including any optimizations made to the code. Finally, it would be helpful to include any test results or benchmarks that demonstrate the improvement in data reading performance.

**Review by m-schen — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

**Review by heni02 — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Pull request merged**
